ENH: option for groupby.hist to match bins

javadnoorb opened the pull request:
- [x] closes #22222
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Example code and result:

```python
%matplotlib inline
import numpy as np
import pandas as pd

def test_hist_bins_match():
    # GH-22222
    N = 100
    bins = 5

    np.random.seed(0)
    df = pd.DataFrame(np.append(np.random.randn(N), np.random.randn(N) / 10),
                   columns=['rand'])
    df['group'] = [0] * N + [1] * N
    g = df.groupby('group')['rand']
    bin_range = np.linspace(df['rand'].min(), df['rand'].max(), bins + 1)

    ax = g.hist(bins=bins, alpha=0.7, equal_bins=True)
    both_hists_max = g.apply(lambda x: max(
        np.histogram(x, bins=bin_range)[0])).max()
    assert ax.iloc[0].get_ylim()[1] >= both_hists_max

test_hist_bins_match()
```

![image](https://user-images.githubusercontent.com/26352146/43753017-add30f56-99d1-11e8-9b20-fecbdfa20fb2.png)

javadnoorb committed:
ENH: option to equalize histogram bin widths

javadnoorb committed:
ENH: option to equalize histogram bin widths

javadnoorb committed:
ENH: option to equalize histogram bin widths

javadnoorb committed:
ENH: fix docstring

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -582,6 +582,15 @@ def wrapper(*args, **kwargs):
                 kwargs_with_axis['axis'] = self.axis
 
+            if (name == 'hist' and
+                    kwargs.pop('equal_bins', False)):
```
Can you get the same behavior without mutating the kwargs dict?

javadnoorb replied:
Yes, I can pass `equal_bins` to `hist_series`. Do you also mean this for line 588? Just asking, because that might be a bit tricky to pipe through.

WillAyd replied:
Less concerned about 588 since you end up replacing that anyway

javadnoorb replied:
Thanks. I ended up copying kwargs and mutating that.

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -582,6 +582,15 @@ def wrapper(*args, **kwargs):
+                    kwargs.pop('equal_bins', False)):
+                # GH-22222
+                bins = kwargs.pop('bins', None)
+                if type(bins) == int:
```
Would be preferable I think to use `is_integer` from `pandas.core.dtypes.inference`

javadnoorb replied:
done

WillAyd commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -12,6 +12,25 @@
+    ax = g.hist(bins=bins, alpha=0.7, equal_bins=True)
+    both_hists_max = g.apply(lambda x: max(
+        np.histogram(x, bins=bin_range)[0])).max()
+    assert ax.iloc[0].get_ylim()[1] >= both_hists_max
```
Is it not possible to make a stronger assertion here about the actual binned values?

javadnoorb replied:
Good suggestion! I'm comparing bar widths in the new test.

WillAyd reviewed (changes requested):
Thanks for the PR!

javadnoorb committed:
ENH/TST/DOC: no mut kwargs - use is_integer/test bar widths/move doc

javadnoorb commented on pandas/plotting/_core.py:
```diff
@@ -2470,8 +2470,11 @@ def hist_series(self, by=None, ax=None, grid=True, xlabelsize=None,
         bin edges, including left edge of first bin and right edge of last
         bin. In this case, bins is returned unmodified.
-    bins: integer, default 10
-        Number of histogram bins to be used
```
This is repeated and seems redundant.

WillAyd replied:
Repeated where? This should still stay, no?

javadnoorb replied:
`bins` is defined on line 2468 and 2473. It looks like a mistake when writing the docstring. What do you think?

WillAyd replied:
Ah OK didn't see that. This is fine then

javadnoorb committed:
TST: more rigorous test

WillAyd commented on pandas/plotting/_core.py:
```diff
@@ -2480,6 +2483,7 @@ def hist_series(self, by=None, ax=None, grid=True, xlabelsize=None,
     matplotlib.axes.Axes.hist : Plot a histogram using matplotlib.
 
     """
+    # TODO: separate docstrings of series and groupby hist functions (GH-22241)
```
Don't need this comment here - the open issue should suffice

WillAyd commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -12,6 +12,23 @@
+    g = df.groupby('group')['rand']
+    ax = g.hist(bins=bins, alpha=0.7, equal_bins=True)[0]
+    bin_width_group0 = ax.patches[0].get_width()
+    bin_width_group1 = ax.patches[bins].get_width()
```
Instead of using `np.random` can you not define a set of data whereby you can also easily assert that both groups have an equal number of bins across the same X-scale?

javadnoorb replied:
I can replace `np.random` with some predefined values. But I think that comparing the number of bins wouldn't do the job. That's the current state of the code. We're here just changing the range without changing the number of bins.

I've written a new test, that compares the x-axis values of the bins, that should be more rigorous. For some reason the functions see kwargs differently in python 2 and 3, so some CI tests fail in the former. I'll make a commit once I figure out what's going on.

WillAyd commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -12,6 +12,23 @@
 
 
+@td.skip_if_no_mpl
+def test_hist_bins_match():
```
Move this function into the class that already exists in the module

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -578,15 +579,27 @@ def wrapper(*args, **kwargs):
                kwargs_with_axis['axis'] is None:
                 kwargs_with_axis['axis'] = self.axis
 
+            if name == 'hist' and kwargs_wo_axis.pop('equal_bins', False):
```
I don't think another copy of kwargs is the cleanest solution here - can you not just `get` from the kwargs dict instead of popping?

javadnoorb replied:
In that case `matplotltib.pyplot.hist` throws an error that `equal_bins` is not recognized. An alternative solution that I can think of is to pass `equal_bins` as a named argument to `hist_series` in pandas/plotting/_core.py. But then it will be a dummy variable that's not used within the function. Any suggestions?

javadnoorb replied:
To clarify a bit more, in python3 (but for some reason not python2), any extra argument passed to `kwargs`will be read in this line (as `kwds`):

https://github.com/pandas-dev/pandas/blob/0370740034978d3a63d4b8e5e2c96ff54e7e08ba/pandas/plotting/_core.py#L2501

javadnoorb commented:
There is a python2 versus python3 discrepancy that's causing the some CI tests to fail. I opened #22285 to deal with it.

javadnoorb committed:
debugging py2/3 discrepancy

javadnoorb committed:
remove kwargs copy mutation/make test more rigorous

javadnoorb committed:
sync

Merge branch 'master' of https://github.com/pandas-dev/pandas into hist_match_bins_tmp1_rebased

javadnoorb committed:
syncing the fork

javadnoorb committed:
more rigorous test / remove kwargs copy

javadnoorb committed:
whitespace fix

javadnoorb commented:
One job failed for no reason. Otherwise it's green.

javadnoorb commented on pandas/plotting/_core.py:
```diff
@@ -2443,7 +2443,7 @@ def hist_frame(data, column=None, by=None, grid=True, xlabelsize=None,
 def hist_series(self, by=None, ax=None, grid=True, xlabelsize=None,
                 xrot=None, ylabelsize=None, yrot=None, figsize=None,
-                bins=10, **kwds):
+                bins=10, equal_bins=False, **kwds):
```
I'm trying the alternative approach here. `equal_bins` will be unused within the function. This is just to avoid popping `kwds`.

WillAyd replied:
Might have missed the compat issue but if you are using `.get` it shouldn't matter whether this exists as a keyword or not

javadnoorb replied:
`kwds` can only contain `matploltib.pyplot.hist` input args (b/c of line 2504) and throws an error if `equal_bins` is a keyword.

javadnoorb commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -10,10 +10,54 @@
+    @pytest.mark.parametrize(
+        'bins, equal_bins',
+        zip([5, None, np.linspace(-3, 3, 10)], [True, False])
+    )
```
Trying all possible forms of `bins` with and without `equal_bins`

WillAyd replied:
Would still prefer if you could be explicit about what is expected here, i.e. add `expected` as a third parameter here and assert that we get the correct number of bins across the two groups

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -581,6 +582,17 @@ def wrapper(*args, **kwargs):
                kwargs_with_axis['axis'] is None:
                 kwargs_with_axis['axis'] = self.axis
+            if (name == 'hist' and
+                    kwargs.get('equal_bins', False) is True):
```
Do we need `is True` here or can we use the implicit truthiness of objects?

javadnoorb replied:
It's just a bit safer this way, as it guards against some non-boolean statements by the user, such as `equal_bins=None`. But your suggestion should be fine if we ignore bad inputs from user.

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -581,6 +582,17 @@ def wrapper(*args, **kwargs):
+                bins = kwargs.get('bins')
+                if bins is None:
+                    bins = 10  # use default value used in `hist_series`
+                if is_integer(bins):
```
Could just use `is_scalar` here since we only need to guard against scalar vs sequence

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -581,6 +582,17 @@ def wrapper(*args, **kwargs):
+                    kwargs.get('equal_bins', False) is True):
+                # GH-22222
+                bins = kwargs.get('bins')
+                if bins is None:
```
What is the purpose of this? Doesn't the keyword already default to 10?

javadnoorb replied:
The keyword defaults to 10 after the function is called (`curried_axis`), but we need its value before that (on line 593)

WillAyd commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -10,10 +10,54 @@
 
 @td.skip_if_no_mpl
 class TestDataFrameGroupByPlots(TestPlotBase):
+    @pytest.mark.parametrize(
```
Readability nit but let's add an empty line here

WillAyd commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -10,10 +10,54 @@
+                index=[0, 1], columns=['min', 'max'])
+            group_ranges = g.agg([min, max])
+            assert np.isclose(group_ranges, hist_ranges).all()
+        tm.close()
```
Not sure what the purpose of this call is?

WillAyd commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -10,10 +10,54 @@
+        axes = g.hist(bins=bins, alpha=0.7, equal_bins=equal_bins)
+        ax = axes[0]
+
+        if bins is None:
```
Just generally avoid this type of code in tests - you aren't asserting anything about the functionality of the code being tested but rather just adding separate logic here in the test case

javadnoorb replied:
The test should be slightly different, depending on the type provided for `bins`. Do you recommend dropping `parametrize` decorator and writing separate tests for these instead?

javadnoorb committed:
CLN: use expected in test / other minor changes

javadnoorb committed:
whatsnew update

javadnoorb commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -640,6 +640,7 @@ Plotting
 
 - Bug in :func:`DataFrame.plot.scatter` and :func:`DataFrame.plot.hexbin` caused x-axis label and ticklabels to disappear when colorbar was on in IPython inline backend (:issue:`10611`, :issue:`10678`, and :issue:`20455`)
 - Bug in plotting a Series with datetimes using :func:`matplotlib.axes.Axes.scatter` (:issue:`22039`)
+- The new argument ``equal_bins`` in :func:`SeriesGroupBy.hist` sets histogram bins to equal values (:issue:`22222`)
```
This is actually `hist` method of a `SeriesGroupBy` object, although that't not available in the docs:

https://pandas.pydata.org/pandas-docs/stable/api.html

javadnoorb replied:
Should it be added?

javadnoorb replied:
This is available in `api.rst` if that suffices:

```
The following methods are available in both ``SeriesGroupBy`` and
``DataFrameGroupBy`` objects, but may differ slightly, usually in that
the ``DataFrameGroupBy`` version usually permits the specification of an
axis argument, and often an argument indicating whether to restrict
application to columns of a specific data type.

.. autosummary::
   :toctree: generated/

...
   DataFrameGroupBy.hist
```

WillAyd replied:
Just change the reference to DataFrameGroupBy.hist in the whatsnew - should suffice

javadnoorb commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -10,11 +10,52 @@
 @td.skip_if_no_mpl
 class TestDataFrameGroupByPlots(TestPlotBase):
 
+    @pytest.mark.parametrize('equal_bins, bins, expected1, expected2', (
```
Need two `expected` variables for the two groups. They correspond to the leftmost part of each histogram bar on the x-axis.

javadnoorb commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -10,11 +10,52 @@
+
+        tm.assert_almost_equal(points[:num_bins], np.array(expected1))
+        tm.assert_almost_equal(points[num_bins:], np.array(expected2))
+        tm.close()
```
this is to close the figure (to avoid overwriting it during `parametrize`)

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -581,6 +581,16 @@ def wrapper(*args, **kwargs):
                 kwargs_with_axis['axis'] = self.axis
+            if name == 'hist' and kwargs.get('equal_bins', False):
+                # GH-22222
+                bins = kwargs.get('bins')
```
Hmm wondering if we even need to use .get here - can we just access by index? I'm assuming that 'equal_bins' should always be there so a KeyError when it's not would be fine

javadnoorb replied:
same py2/3 issue. In py3 `equal_bins`, `bins`, etc do not take their default values unless the code inside `hist_series` function is being run.

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -581,6 +581,16 @@ def wrapper(*args, **kwargs):
             if 'axis' not in kwargs_with_axis or \
                kwargs_with_axis['axis'] is None:
                 kwargs_with_axis['axis'] = self.axis
+            if name == 'hist' and kwargs.get('equal_bins', False):
```
Can just do `kwargs.get('equal_bins')` - no need for the `False` argument

javadnoorb replied:
For some reason py3 and py2 `kwargs` behavior are different. In py3 `kwargs` defaults are not set here if the parameter is not provided by the user. I opened an issue on this (#22285). I'm not sure where to move the code (per @jreback's comment) to avoid this.

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -581,6 +581,16 @@ def wrapper(*args, **kwargs):
                 kwargs_with_axis['axis'] = self.axis
+            if name == 'hist' and kwargs.get('equal_bins', False):
+                # GH-22222
+                bins = kwargs.get('bins')
```
do not  add anything here

this is obscuring already extremely fragile code

javadnoorb replied:
Do you mean not in the wrapper? Or also beyond that? Is there any other place you'd suggest?

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -581,6 +581,16 @@ def wrapper(*args, **kwargs):
+            if name == 'hist' and kwargs.get('equal_bins', False):
+                # GH-22222
+                bins = kwargs.get('bins')
+                if bins is None:
```
Can you point me to where `bins` doesn't default to 10?

WillAyd commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -10,11 +10,52 @@
+        (True, 10,
+         [-3., -2.4, -1.8, -1.2, -0.6, 0., 0.6, 1.2, 1.8, 2.4],
+         [-3., -2.4, -1.8, -1.2, -0.6, 0., 0.6, 1.2, 1.8, 2.4]),
+        (True, None,
```
Assuming the default is always 10 for `bins` then `None` is not a valid argument here and should in fact raise an error instead of giving valid results

WillAyd commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -10,11 +10,52 @@
+        points = np.array([patch.get_bbox().get_points()
+                           for patch in ax.patches])[:, 0, 0]
+
+        tm.assert_almost_equal(points[:num_bins], np.array(expected1))
```
Would prefer to use `tm.assert_numpy_array_equal` instead here and on line below

javadnoorb commented:
Making a new function that does this is pretty straightforward:
```diff
diff --git a/pandas/core/groupby/groupby.py b/pandas/core/groupby/groupby.py
index 3f84fa0..2d324f4 100644
--- a/pandas/core/groupby/groupby.py
+++ b/pandas/core/groupby/groupby.py
@@ -1001,6 +1001,13 @@ class GroupBy(_GroupBy):
     len(grouped) : int
         Number of groups
     """
+    def new_hist(self, bins=10, equal_bins=False, **kwargs):
+        if equal_bins and is_scalar(bins):
+            # share the same numpy array for all group bins
+            bins = np.linspace(self.obj.min(),
+                               self.obj.max(), bins + 1)
+        return self.hist(bins=bins, **kwargs)
+
     def _bool_agg(self, val_test, skipna):
         """Shared func to call any / all Cython GroupBy implementations"""
```

but I don't know how to overload `SeriesGroupBy.hist` without touching the wrappers, or some other sensitive part of the code. I defer this PR to anyone who knows how to do this.

WillAyd commented:
Closing as stale. Ping if you'd like to pick this back up
